pkgs/stdenv/generic/make-derivation.nix

```diff
@@ -354,7 +363,14 @@ in rec {
       lib.extendDerivation
         validity.handled
         ({
-           overrideAttrs = f: mkDerivation (attrs // (f attrs));
+           overrideAttrs = f:
+              remakeDerivation (
+                  # remakeDerivation will pass us the original function
+                  oldf:
+                    # this function body is exactly like a mkDerivation argument
+                    { this, ... }@t:
+                    let attrs = oldf t; in attrs // f (attrs // { inherit this; })
```

Comment:
f's parameter is like `super`, but what if we did another round of `builtins.isFunction` trickery so we could have `super` and `self` parameters? Then this would better match other such things in Nixpkgs, and also, we can get rid off `rec { ... }` `mkDerivation` arguments which are hard to override.

I think we can also use `lib.fixed-points.extend` at that point.

Reply:
:100: 

I thought of doing `self: super:` but considered it too ugly. I didn't quite realize we could do reflection inside the partially applied function cheaply.

> we can get rid off rec { ... } mkDerivation arguments which are hard to override.

This is definitely a possibility, even before the refactor, but I got an objection about this (big) change earlier. Whether or not the community wants to change "style" is not in any way dictated by this PR though.